# 예외 처리

자바에서는 컴파일하여 *.class 파일을 생성한 후 실행을 할 수 있는데, 컴파일이 성공했다고 해서 프로그램 실행시에도 에러가 발생하지 않는다고 할 수는 없다.
실행 도중에 에러가 발생하여 프로그램이 예기치 못한 상태가 되는것을 대비하기 위하여 예외 처리를 사용한다.


## 체크 예외 와 언체크 예외

**체크 예외** 
Exception 클래스의 서브클래스이면서 RuntimeException 클래스를 상속하지 않은 것들이다.
체크예외가 발생할 수 있는 메소드를 사용할 경우 반드시 예외를 처리하는 코드를 함께 작성해야 한다.
그렇지 않으면 컴파일 에러가 발생한다.


**언체크 예외** 
RuntimeException을 상속한 클래스이다. RuntimeException 클래스를 상속한 예외들은 예외처리를 강제하지 않기 때문에 언체크 예외, 런타임 예외라고 불린다.
런타임 예외는 프로그램에 오류가 있을 때 발생하도록 의도된 것들이다. 이런 예외에 대한 부분들을 개발자가 미리 체크하고 처리하면 피할 수 있지만,
아직 처리하지 못한 부분에 대해서 발생하도록 만든 것이 런타임 예외이다.


**결국 Exception은 두가지 분류로 나눌 수 있다.**
1. Exception 클래스와 Exception 클래스를 상속받은 나머지 클래스 ( RuntimeException 제외, 컴파일 시 발생 )
2. RuntimeException 클래스와 RuntimeException 클래스를 상속받은 클래스 ( 실행 시 발생 )

나누는 기준은 컴파일시 예외 처리 체크를 하는지 안하는지에 있다. 컴파일러는 RuntimeException을 제외한 모든 Exception 클래스들을 컴파일 시 예외 처리( try-catch )를 했는지
반드시 확인한다.

**왜 RuntimeException은 try-catch가 사용되었는지 확인을 안할까?**
RuntimeException에 속하는 클래스들은 주로 개발자의 실수에 의해 발생되는 예외들이 많기 때문이다.
예를들어 NullPointerException(널 객체를 참조했을 경우), ArrayIndexOutOfBoundsException(배열을 참조하는 인덱스가 잘못된 경우) 등이 있다.
이런 예외들은 개발자가 좀 더 신경쓰면 발생하지 않게 할 수 있으므로 컴파일러가 확인하지 않는다고 볼 수 있다.


## 예외 넘기기( throws )
메소드 내부에서 예외가 발생할 수 있는 부분에 try-catch로 예외를 처리하지만, 때로는 throws를 사용하여 호출한 곳으로 예외를 넘길 수도 있다.
throws가 붙어있는 메소드는 반드시 try 블록 내에서 호출 되어야 하고 catch 블록에서 받은 예외를 처리해야 한다.


```java
public void method1(){
    try{
        method2();
    }catch(Exception e){
        System.out.println("클래스가 존재하지 않습니다.");
    }
}


public void method2() throws Exception{
    
}
```